latch/services/cp.py

```diff
@@ -54,71 +57,84 @@ def _cp_local_to_remote(local_source: str, remote_dest: str):
-        raise ValueError(
-            "your token has expired - please run latch login to refresh your token and try again."
-        )
+    token = retrieve_or_login()
+    if local_source.is_dir():
+        mkdir(remote_directory=remote_dest)
+        tasks = []
+        for sub_dir in local_source.iterdir():
+            tasks.append(
+                threading.Thread(
+                    target=_cp_local_to_remote,
+                    kwargs={
+                        "local_source": sub_dir,
+                        "remote_dest": f"{remote_dest}/{sub_dir.name}",
+                    },
+                )
+            )
+
+        for task in tasks:
+            task.start()
+
+        for task in tasks:
+            task.join()
 
-    r_json = response.json()
-    path = r_json["path"]
-    upload_id = r_json["upload_id"]
-    urls = r_json["urls"]
-
-    parts = []
-    print(f"\t{local_source.name} -> {remote_dest}")
-    total_mb = total_bytes // 1000000
-    for i in range(nrof_parts):
-
-        if i < nrof_parts - 1:
-            _end_char = "\r"
-        else:
-            _end_char = "\n"
-
-        print(
-            f"\t\tcopying part {i+1}/{nrof_parts} ~ {min(total_mb, (_CHUNK_SIZE//1000000)*(i+1))}MB/{total_mb}MB",
-            end=_end_char,
-            flush=True,
-        )
-        url = urls[str(i)]
+    else:
         with open(local_source, "rb") as f:
-            f.seek(i * _CHUNK_SIZE, 0)
-            resp = requests.put(url, f.read(_CHUNK_SIZE))
-            etag = resp.headers["ETag"]
-            parts.append({"ETag": etag, "PartNumber": i + 1})
-
-    data = {"path": path, "upload_id": upload_id, "parts": parts}
-    url = endpoints["complete-multipart-upload"]
-    headers = {"Authorization": f"Bearer {token}"}
-    response = requests.post(url, headers=headers, json=data)
+            f.seek(0, 2)
+            total_bytes = f.tell()
+            f.seek(0, 0)
+
+        nrof_parts = math.ceil(total_bytes / _CHUNK_SIZE)
+
+        data = {
+            "dest_path": remote_dest,
+            "node_name": local_source.name,
+            "content_type": "text/plain",
+            "nrof_parts": nrof_parts,
+        }
+        url = endpoints["initiate-multipart-upload"]
+        headers = {"Authorization": f"Bearer {token}"}
+        response = requests.post(url, headers=headers, json=data)
+
+        print(response.content)
+
+        r_json = response.json()
+        path = r_json["path"]
+        upload_id = r_json["upload_id"]
+        urls = r_json["urls"]
+
+        parts = []
+        print(f"\t{local_source.name} -> {remote_dest}")
+        total_mb = total_bytes // 1000000
+        for i in range(nrof_parts):
+
+            if i < nrof_parts - 1:
+                _end_char = "\r"
+            else:
+                _end_char = "\n"
+
+            print(
+                f"\t\tcopying part {i+1}/{nrof_parts} ~ {min(total_mb, (_CHUNK_SIZE//1000000)*(i+1))}MB/{total_mb}MB",
```

Comment:
i know this is based off of what I did, but this will quickly look bad

eg. 10 KB or 10GB expressed as MB instead of nearest human readable unit

consider writing a simple pretty printer
https://stackoverflow.com/questions/10420352/converting-file-size-in-bytes-to-human-readable-string